### ISSUE TITLE
Filter log lines by Regexp before they are sent

### ISSUE THREAD

**Issue by ruflin** [elastic/filebeat#78]
A configuration option to define a regexp per prospector to filter each log line could be introduced.

This was requested for logstash-forwarder: https://github.com/elastic/logstash-forwarder/issues/482

**Comment by ruflin** [comment]
An interesting use case is to filter usernames/password/hashes which were log and should be filtered out before sent to the server.

**Comment by monicasarbu** [comment]
In Packetbeat, we hide the passwords from the HTTP messages before sending the transaction to the output. Here are details:
https://www.elastic.co/guide/en/beats/packetbeat/current/packetbeat-configuration.html#_http_configuration

We should make this feature available for all Beats and be part of filtering inputs.

**Comment by roji** [comment]
Seems to be that filtering out sensitive info from lines is a different feature from this issue, which is about filtering out the entire log record (i.e. not shipping it). Maybe a separate issue?

**Comment by sastorsl** [comment]
Would really like to see this. It should be possible to split logs into different types, dropping certain keywords, etc.
Think log files with different input, i.e. maillog (postfix and imap), syslog (filter out arbitrary "logger" input).

I'm having issues with a maillog where uninteresting imapd LOGIN/LOGOUT floods logstash, while I would like to keep only postfix logs. Being able to filter these at the shipper would be great.

**Comment by monicasarbu** [comment]
@roji Yes, true. It's a different use case, but we are planning to implement a "generic" filtering algorithm that can decide what log message to drop or change before sending it to Logstash. We are still in the phase of collecting ideas about what the filtering algorithm should do. So, feel free to give us more inputs about it.

**Comment by ruflin** [comment]
As reference, see https://github.com/elastic/logstash-forwarder/pull/208 for a potential implementation in lsf

**Comment by andrewkroh** [comment]
Winlogbeat will also need to implement filtering (i.e. filtering by time, event ID, severity, user, category, etc.). One design consideration I wanted to bring up is about where the filtering occurs. If we have a filter chain it would be ideal if some of the filters could be executed as close to the data source as possible to avoid unnecessary processing. One trade-off is that it does introduce extra complexity.

For example, in Winlogbeat if there is a filter applied to only allow event ID 7042, then there is no need to waste resources fully processing messages with other event IDs (but the beat does need to inform the registrar that the event was read so that it can keep track of the current read location).

**Comment by andrewkroh** [comment]
Winlogbeat will also need to implement filtering (i.e. filtering by time, event ID, severity, user, category, etc.). One design consideration I wanted to bring up is about where the filtering occurs. If we have a filter chain it would be ideal if some of the filters could be executed as close to the data source as possible to avoid unnecessary processing. One trade-off is that it does introduce extra complexity.

For example, in Winlogbeat if there is a filter applied to only allow event ID 7042, then there is no need to waste resources fully processing messages with other event IDs (but the beat does need to inform the registrar that the event was read so that it can keep track of the current read location).

**Comment by monicasarbu** [comment]
I propose to add an option in Filebeat to ignore the log lines that match a certain regexp.
To ignore all the DBG lines you can use:

<pre>
ignore_lines: "^DBG"
</pre>

**Comment by ruflin** [comment]
I'm very undecided if we should go with an "include" or "exclude" pattern. The advantage of going with an exclude pattern as @monicasarbu suggests is that it is specified very clear what not should be sent. On the other hand this PR suggest the opposite implementation: https://github.com/elastic/logstash-forwarder/pull/208

@monicasarbu suggestion is probably better as in case of doubt, it will send the log line. Better to have one log line too much, then one too few.

Should we call it `filter: "^DBG"' instead? I'm fine with both options as also `ignore_lines` is very descriptive on what it does.

**Comment by tsg** [comment]
+1 for the ignore version. But I think we would need it to accept an array of regexps, because it will be common for people to exclude more than one pattern.

**Comment by ruflin** [comment]
@tsg Should we allow an array or use `^DBG|^WARN`?

**Comment by monicasarbu** [comment]
@ruflin I will avoid to use the "filter" name for it, not to confuse people with filters: https://github.com/elastic/libbeat/issues/336.

@tsg I like the idea of having a list of regexp, it's easier to implement than "or" or "and".

**Comment by ruflin** [comment]
@monicasarbu I would only go with OR in the beginning.

**Comment by monicasarbu** [comment]
If it's requested than I can also implement allow_lines: ["^ERR", "^WANR"].

**Comment by ruflin** [comment]
Lets go with your simple suggestion above first with the array suggested by @tsg and interate on top of it.

**Comment by tsg** [comment]
@ruflin `^DBG|^WARN` might be a good idea. I cannot think of a case where this won't work, just that it might result in really  long (and maybe slow?) regexps. I'm fine with both options.

**Comment by ruflin** [comment]
@tsg yours is probably more "readable". +1 on yours.

**Comment by urso** [comment]
I think it can make sense to have both 'include_lines' and 'exclude_lines'. Plus have both receive a list of patterns. Thinking on how often I've chained grep commands to simulate this behavior on syslog files I would expect this to be a must have.

**Comment by urso** [comment]
I think golang regexp supports posix regular expressions. So '|' should be supported already: http://play.golang.org/p/L1zuN-yKhq

**Comment by mjmeyer** [comment]
+1 for this feature and suggest that if only one idiom is to be  supported that it be the include form. 

I believe that the use case supporting this in along the lines of:
- I want efficient shipping of log data to remote logstash...hence choosing filebeat
- I dont wish to ship the entire log
- log data is fuzzy by nature. I know of specific pattern(s) of lines I want shipped but dont know EVERYTHING that is written to log today or especially in future.

**Comment by roji** [comment]
+1 for this, but please include both include and exclude. It's an extremely common scenario to have a log where everything interests us except for a few line types which are repeated a lot (e.g. some performance stat). It's important to be able to filter those out.

**Comment by sastorsl** [comment]
+1 This is what I'm interested in too.
However, with both include and exclude, what would you expect to take precedence?

Include everything in include.
Exclude everything in exclude.

And if both are present, what would you expect to happen?

In accordance with stringing along pipes with "grep" or if-then-else in awk (or similar) I would expect that I _could_ apply advanced pattern matches.

Use case1: Only Include, no exclude = ship only lines matching include.
Use case2: Only exclude, no include = ship all lines, exclude matching patterns.
Use case3: Include first, exclude second = Include matching patterns _except_ exclude matches
Use case4: Exclude first, include second = ship all except Exclude patterns, but include matches in the exclude pattern which match the include pattern.

Using yaml or json it should be possible to make this even more advanced.
But use case 1 and 2 is IMHO a priority.

**Comment by roji** [comment]
The most complete/complex solution here would be a firewall-like ordered list of rules, each one being a regex and an action (drop or keep). This might be an overkill but would allow all scenarios.

Otherwise something like what @sastorsl proposes could be a simpler alternative.

**Comment by monicasarbu** [comment]
@sastorsl @roji @mjmeyer 
Thank you everyone for the feedback. I am working on a pull request that adds the `include_lines` and `exclude_lines` options. In the current implementation, the first three use cases described are covered:

- `exclude_lines` option accepts a list of regular expressions to match. If any line matches at least one of the regular expressions, the line is dropped and not exported
- `include_lines` option accepts a list of regular expressions to match. If any line matches at least one of the regular expressions, the line is exported

If both options are defined, then include_lines is executed first. 

__Examples:__

Include only the error and warning log messages:
<pre>
include_lines: ["^ERR", "^WARN"]
</pre>

Exclude the debug messages:
<pre>
exclude_lines: ["^DBG"]
</pre>

Include the apache logs, but without the debug messages:
<pre>
include_lines: ["apache"]
exclude_lines: ["^DBG"]
</pre>

I can try to add another option that defines which rule to be executed first, include_lines or exclude_lines. Do you guys think Use case4 is commonly used?

**Comment by roji** [comment]
@monicasarbu usecase4 seems pretty rare, your plan looks good to me.

**Comment by sastorsl** [comment]
Use case4: "exclude imap except on ERROR" --> exclude: imap, include imap ERROR

I would settle for the possibility to do this, while I want to emphasize that covering the basic include and exclude patterns should be a priority, and simple for most people.

The more complex patterns should be possible and documented but I would expect more from people trying to use them so they can be complicated to implement for the user.

**Comment by monicasarbu** [comment]
@sastorsl For these exceptional cases, we are planning to implement filtering in all Beats. Please have a look about more details here: https://github.com/elastic/beats/issues/451. Do you think people will use the generic filtering to implement the use case 4? or do you think they will prefer defining another option in the configuration file?

**Comment by sastorsl** [comment]
Yes, that would be a viable approach. However, I depend on logstash-forwarder to add fields, mainly "type" and "IndexType". Not sure if I understand your proposal correctly, but will filters give that possibility?

Example:

```json
  # logstash-forwarder config
  "files": [
    {
      "paths": [
        ".../logs/serviceone_a-access_log",
        ".../logs/serviceone_b-access_log"
       ],
      "fields": { "type": "comp_apache_special", "IndexType": "serviceone-http" }
    }, {
      "paths": [
        ".../logs/serviceone_c-access_log"
       ],
      "fields": { "type": "comp_apache_specialcookie", "IndexType": "serviceone-http" }
    }, {
      "paths": [
        ".../logs/servicetwo_access_log"
       ],
      "fields": { "type": "comp_apache_specialcookie", "IndexType": "servicetwo-http" }
    }
  ]
}
```

**Comment by monicasarbu** [comment]
@sastorsl Thank you for your feedback. In the first phase, we are planning to add only `drop_event` and `drop_fields`. Could you please add your request to the discussion elastic/beats#451? Thanks!

**Comment by monicasarbu** [comment]
@sastorsl Currently, you can add these fields by using the `fields` option. 
<pre>
fields:
    type: comp_apache_special
    IndexType: serviceone-http
</pre>

The main purpose of the filtering is to remove data from your event, not adding additional fields. I think, it makes more sense to keep adding fields with the option `fields`.

**Comment by sastorsl** [comment]
Ah, yes. Thank you for your clarification. Agreed.

**Comment by ruflin** [comment]
@sastorsl One small addition: In case you just want to set the document type, you can do that for each prospector through the document_type config option: https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-configuration-details.html#_document_type

**Comment by monicasarbu** [comment]
Closing this issue as it is implemented in the PR: https://github.com/elastic/beats/pull/430 and it will be available in the Beats 1.0.1 release.
